### ISSUE TITLE
Copy notification improvements

### ISSUE THREAD

**Issue by bender2k14** [CONNECT-platform/codedoc#57]
To be precise, I will use some [UX terminology](https://ux.stackexchange.com/questions/103783/success-toast-vs-modal).
- Modal popup: a popup that requires a user action to dismiss.
- Toast popup: a popup that disappears after a short time without user interaction.

When a user copies something using the links built into a CODEDOC site, the user is notified via a full-screen modal popup.  For example, here is a gif of me copying the URL to a heading on my Android phone.

![ezgif-6-e9c8e0aaf9a4](https://user-images.githubusercontent.com/34664007/88469151-b63adc00-ceb3-11ea-917d-4e1ed3e8f3d8.gif)

As part of [this UX.SE answer says](https://ux.stackexchange.com/questions/12045/what-is-a-modal-dialog-window/12049#12049),
> Modal dialogs should be used when the user takes an action that needs additional information to be completed correctly. The print dialog, is also a good example. Or the 'options' dialogs in loads of programs.
> 
> When to use modal dialogs is a bit tricky to define, but to me it should be used sparingly, and avoided whenever possible.

The operation of copying the heading URL doesn't require information from the user, so according to this answer, it shouldn't use a modal popup.  I would prefer if this copy notification was achieved via a toast popup.

In fact, copying something to the clip tray already produces a toast popup on Android.  In this case, the additional popup provided by CODEDOC is somewhat redundant.

It is not completely redundant because the popups contain different information.  However, I prefer the brevity of the Android popup.  It says that something was copied but doesn't overload me with the precise text that was copied.  Something that might be better than both of these is a popup that says something like "Heading link copied" (which conveys at a high level what was copied but not exactly what was copied), but I am not sure about that.

I also like that the Android popup only covers a small part of the screen.  The full-screen nature of the CODEDOC popup steals my focus and breaks my [flow](https://en.wikipedia.org/wiki/Flow_(psychology)).

I am not a UX expert, so this post is my best attempt to share how I feel when interacting with this CODEDOC feature and what I think a UX expert might have to say about it.

Are improvements to the CODEDOC copy notification something that can be considered?

**Comment by loreanvictor** [comment]
> Are improvements to the CODEDOC copy notification something that can be considered?

Of course, like all other aspects.

<br><br>

> I also like that the Android popup only covers a small part of the screen. The full-screen nature of the CODEDOC popup steals my focus and breaks my flow.

May I ask what is the thing you are focused on when you copy something, and how the popup steals that focus?
My assumption on that was:

- When you copy something, the next thing you want to do is pasting it.
- Which means the immediate action after copying is typically not resuming reading, but its rather navigating away.
- Which means blurring the text you are reading shouldn't break your flow (since your flow typically doesn't involve focus on the information that is losing focus).

<br><br>

> In fact, copying something to the clip tray already produces a toast popup on Android. In this case, the additional popup provided by CODEDOC is somewhat redundant.

To be completely honest I haven't checked out the consistency of existence of such toasts, but can say that it would at least entail checking the platform and showing or not showing a confirmation depending on whether or not the platform itself would do that. This is in addition to the fact that the toast almost never shows exactly what was the copied text, which is a piece of information I really intended on communicating via the copy confirmation overlay.

<br><br>

> The operation of copying the heading URL doesn't require information from the user, so according to this answer, it shouldn't use a modal popup. I would prefer if this copy notification was achieved via a toast popup.

To reference the answer in the [first link you mentioned](https://ux.stackexchange.com/questions/103783/success-toast-vs-modal):

> Generally, modals should be avoided for feedback of operations. They are "Ok" when a review of the task is needed or some other manual validation.

And that is precisely the intent here. Not just to send a "confirm" message and move on, but to provide a chance for the user to _review_ what has been copied. This is the backdrop to this whole interaction: I was always annoyed that I didn't get proper confirmation when I copied code / text / links via `click-to-copy` mechanisms, to the extent that if I wanted to copy the text somewhere important that might get accidentally sent (sending a message to a coworker) then I would first paste the link into the address bar of the browser and check it.

Now assuming the aforementioned user flow for copying something (_copy, navigate away, paste_), after the _copy_ step none of the information on the page is useful anymore (since the intent is to navigate away), which means you have all of the screen space at your disposal to convey detailed confirmation of the success of the first step and allow the user to properly review what they have copied.

I of course realize that the feeling of safety driving this design choice varies per individual, but I am curious about how this design breaks the user flow, to try and find a solution that avoids that while still providing same amount of safety / control.

**Comment by bender2k14** [comment]
> May I ask what is the thing you are focused on when you copy something, and how the popup steals that focus?
> My assumption on that was:
> 
> * When you copy something, the next thing you want to do is pasting it.
> * Which means the immediate action after copying is typically not resuming reading, but its rather navigating away.
> * Which means blurring the text you are reading shouldn't break your flow (since your flow typically doesn't involve focus on the information that is losing focus).

Certainly :)

The first use case that comes to mind is when I copy on accident.  The second one is when I want to copy the text in a heading (c.f. #35).  The combined popup and copy behavior makes this impossible.  The third is when I copy something but want to continue reading a bit before pasting elsewhere.  Maybe that last one is a bad workflow on my part.

> To be completely honest I haven't checked out the consistency of existence of such toasts, but can say that it would at least entail checking the platform and showing or not showing a confirmation depending on whether or not the platform itself would do that.

I think this is only a minor issue.  I point it out more as an example of where the response to copying is a toast notification.

> This is in addition to the fact that the toast almost never shows exactly what was the copied text, which is a piece of information I really intended on communicating via the copy confirmation overlay.
> 
> > The operation of copying the heading URL doesn't require information from the user, so according to this answer, it shouldn't use a modal popup. I would prefer if this copy notification was achieved via a toast popup.
> 
> To reference the answer in the [first link you mentioned](https://ux.stackexchange.com/questions/103783/success-toast-vs-modal):
> 
> > Generally, modals should be avoided for feedback of operations. They are "Ok" when a review of the task is needed or some other manual validation.
> 
> And that is precisely the intent here. Not just to send a "confirm" message and move on, but to provide a chance for the user to _review_ what has been copied. This is the backdrop to this whole interaction: I was always annoyed that I didn't get proper confirmation when I copied code / text / links via `click-to-copy` mechanisms, to the extent that if I wanted to copy the text somewhere important that might get accidentally sent (sending a message to a coworker) then I would first paste the link into the address bar of the browser and check it.
> 
> Now assuming the aforementioned user flow for copying something (_copy, navigate away, paste_), after the _copy_ step none of the information on the page is useful anymore (since the intent is to navigate away), which means you have all of the screen space at your disposal to convey detailed confirmation of the success of the first step and allow the user to properly review what they have copied.
> 
> I of course realize that the feeling of safety driving this design choice varies per individual, but I am curious about how this design breaks the user flow, to try and find a solution that avoids that while still providing same amount of safety / control.

I think it is possible to achieve your goal without a full-screen modal popup.  What do you think about this simulation of how it could work?  It is specific to the copying of a heading URL and makes use of the white space under a heading.  This notification is so unobtrusive that I think either modal or toast would work.

![clip-notify](https://user-images.githubusercontent.com/34664007/88482254-bb387380-cf25-11ea-844d-d4f4745efc8e.gif)

As for verifying if a copied URL is as intended, this is normally achieved by automatically navigating to that URL.  For example, this is the behavior on GitHub.

![2020-07-26_10-32-04_199](https://user-images.githubusercontent.com/34664007/88483118-67c92400-cf2b-11ea-9c1c-e0f8b2c384e5.gif)

So CODEDOC currently copies the heading URL when clicked and GitHub navigates to the URL when clicked.  I think having both behaviors at the same time would be even better.

I will have to think more about the the current full-screen modal popup behavior for the notification when copying code.

**Comment by loreanvictor** [comment]
> So CODEDOC currently copies the heading URL when clicked and GitHub navigates to the URL when clicked. I think having both behaviors at the same time would be even better.

GitHub navigating to the page is a bit weird, since the user is already where that link would get them. It merely causes a jump in the content which is a bit unsavory.

<br>

> As for verifying if a copied URL is as intended, this is normally achieved by automatically navigating to that URL. For example, this is the behavior on GitHub.

This doesn't work for safari (the detailed URL is not shown), for mobile (the header is typically hidden due to scrolling), or for long URLs (the addition of the hashtag might easily fall into a cropped part).

<br>

> I think it is possible to achieve your goal without a full-screen modal popup. What do you think about this simulation of how it could work? It is specific to the copying of a heading URL and makes use of the white space under a heading. This notification is so unobtrusive that I think either modal or toast would work.

This solution is neat, however I fear it is not really that visible. The issue is, in most workflows (besides copying something and then reading on without pasting it), after the click, the most important information is the copied content (for confirmation) and not rest of the content of the page, a hierarchy which is not reflected in how the information is displayed in the proposed solution (or how it would be displayed via a toast).

<br>

> The second one is when I want to copy the text in a heading (c.f. [#35](https://github.com/CONNECT-platform/codedoc/issues/35)). The combined popup and copy behavior makes this impossible.

This is a valid use case, though as discussed in #35 , it takes much less priority compared to interaction for copying (in terms of ease of use). I suspect, however, that I can actually distinguish between a drag and a click on headings, making this completely possible without taking away ease of copying headings.

<br>

> The first use case that comes to mind is when I copy on accident.

This is a REALLY valid concern, since it means the document basically overreacts to simple mis-clicks. I suspect the error also happens a lot on mobile phones since users need to touch the screen to scroll. It is a bit of a weird situation since the huge tap zone is there specifically for ease of use for mobile users.

<br>

---

## Proposed Solution:

Since the main issue is with error / copying the heading text, I am going to propose a solution based on these assumptions:

- I assume the error mostly happens on mobile
- I assume the copy heading text is mostly attempted on desktop

Based on them, I propose to:

- Change the tap to a longer-press tap on mobile to trigger copy (to avoid errors with scrolling)
- Change the click detection on desktop so that it distinguishes dragging / selection from clicks

Of course, since long tap is already an (often awkward) interaction on phones (for copying text), I don't know about compatibility of the proposed change for mobile phones. Should research that a bit further.

**Comment by bender2k14** [comment]
> GitHub navigating to the page is a bit weird, since the user is already where that link would get them. It merely causes a jump in the content which is a bit unsavory.

Oh, I thought you would like this behavior since then...

> ...if I wanted to copy the text somewhere important that might get accidentally sent (sending a message to a coworker) then I would first paste the link into the address bar of the browser and check it.

...the act of copying the URL also verifies it for you.  I like this behavior.  I verify every webpage URL I copy before sending elsewhere by navigating to it.  Combining these copy and verify steps saves me time.

> > As for verifying if a copied URL is as intended, this is normally achieved by automatically navigating to that URL. For example, this is the behavior on GitHub.
> 
> This doesn't work for safari (the detailed URL is not shown), for mobile (the header is typically hidden due to scrolling), or for long URLs (the addition of the hashtag might easily fall into a cropped part).

I think I might have been unclear there.  I am not saying that this is achieved by navigating to the URL and then looking at the resulting URL in the address bar.  I am saying that this is achieved by navigating to the URL and looking at the displayed content (i.e. the rendered HTML and CSS).

> The issue is, in most workflows (besides copying something and then reading on without pasting it)...
> 
> Now assuming the aforementioned user flow for copying something (copy, navigate away, paste), after the copy step none of the information on the page is useful anymore (since the intent is to navigate away), which means you have all of the screen space at your disposal to convey detailed confirmation of the success of the first step and allow the user to properly review what they have copied.

My experience does not lead me to believe that this assumption is true.  Here are more uses cases to the contrary.

This framing of the "navigate away" step makes me think it is like "navigate away forever and never come back".  When I was creating https://github.com/CONNECT-platform/coding-blog-plugin/issues/12, the steps I did were
1. copy the URL to `Big`
2. navigate away
3. paste
4. navigate back
5. Click to dismiss the full-screen modal popup
6. copy the URL to `Space`
7. navigate away
8. paste

I would prefer if step 5 were not there.  I wouldn't be surprised if this use case of copying more than one link is my most frequent one.

In another use case, I was sharing my screen with a coworker and directing his attention on my screen by highlighting text on the page.  At one point, I highlighted a heading, which triggered the full-screen modal popup and blocked the text that each of us was reading.

> I of course realize that the feeling of safety driving this design choice varies per individual, but...and find a solution that avoids that while still providing same amount of safety / control.

I don't understand what safety or control is being gained.  Let's separate the discussion by URL or not.

I spend very little time reading any URL that I copy.  The primary audience for URLs are programs like browsers or GitKraken.  People are secondary.  If I have a URL, then I spend most of my time giving it to such programs to see if the URL makes sense to that program...if the program responds in the way that I expect.

If I copy text that is not a URL, then people (especially those like myself) are the primary audience.  I do want to see what I copied.  However, this will necessarily happen when I paste....into a blog post, into Slack, into an email.  Furthermore, the exact formatting from the source might not be very good, so I spend a small amount of time improving it.  All the while, I am implicitly verifying that what I copied is correct.  No explicit verification step is needed.

The only benefit I see to the eager full-screen modal popup approach is when the copied text is wrong.  Then I can try again right away instead of realizing only after navigating away and pasting that the copied text is not correct.  But I rarely copy the wrong thing....and this isn't to say that I am really good at copying.  It is a complement to all websites and programs with interfaces that do a great job conveying what will be copied (without having to shove it in my face).

> I was always annoyed that I didn't get proper confirmation when I copied code / text / links via `click-to-copy` mechanisms, to the extent that if I wanted to copy the text somewhere important that might get accidentally sent (sending a message to a coworker) then I would first paste the link into the address bar of the browser and check it.

Your experience with `click-to-copy` seems different than mine.  I think of `click-to-copy` as a casual interaction between parties with an appropriately high level of trust.  I get the feeling that you were burned by some use of `click-to-copy`.  So now you don't trust any use `click-to-copy`.  If I didn't trust `click-to-copy`, then I would also want to compensate by making the interaction more formal.

I think the design goal should be to keep the interaction casual and the relationship healthy via a high level of trust.

**Comment by loreanvictor** [comment]
> 1. copy the URL to Big
> 2. navigate away
> 3. paste
> 4. navigate back
> 5. Click to dismiss the full-screen modal popup
> 6. copy the URL to Space
> 7. navigate away
> 8. paste

Ok I realize the convenience for step 5 not being there. What I meant though was that after step 1, you don't need the content of the page anymore (until step 4), what you need is confirmed knowledge of what you've copied (what will appear when you paste). This is now a matter of convenience instead of reading flow being broken, since by taking step 5 out the reading flow would be equally broken (per user's intent).

> In another use case, I was sharing my screen with a coworker and directing his attention on my screen by highlighting text on the page. At one point, I highlighted a heading, which triggered the full-screen modal popup and blocked the text that each of us was reading.

Yes I agree this behavior should be fixed (see the proposed solution).

And the rest are good points, I need to read and reflect more thoroughly on them.

**Comment by loreanvictor** [comment]
deployed via `0.2.11`.

**Comment by loreanvictor** [comment]
to sum up what was done for this issue:

- the default behavior is changed to a simple swipe-able toast. in case of copying code, there is a button on the toast to also review the copied code.
- the header's text now will not be copied when trying to select it.
- I found no collision between scrolling and copying interactions while testing on mobile devices.

**Comment by bender2k14** [comment]
Sweet.  I am excited to test this.  I plan to test this and all your other recent improvements this weekend.

**Comment by loreanvictor** [comment]
> Sweet. I am excited to test this. I plan to test this and all your other recent improvements this weekend.

Thanks a lot! If you could, please test it on mobile as well. I haven't programmed swipe behavior before, and am curious about how it fares.

**Comment by bender2k14** [comment]
I love the new behavior!  It works great on my phone too.  I tried touching and scrolling in several different ways, and it only copied when I tried to (by just tapping).

Clicking on a line of code has the new toast notification.  However, clicking on the line number for a line of code still has the modal notification.  Did you forgot to change that?

Here is one minor thing about mobile.  See these two screenshots.

### Before tapping heading
![before](https://user-images.githubusercontent.com/34664007/90302773-5611d700-de6e-11ea-8a98-8e3f88b6f0b2.jpg)

### After tapping heading
![after](https://user-images.githubusercontent.com/34664007/90302777-59a55e00-de6e-11ea-8cb0-e290f1ff7429.jpg)

What you can see there on the left is the a small portion of the right side of the copy icon that appears on mouse hover over a heading when using the desktop.  My guess is that this behavior is not intended.

**Comment by loreanvictor** [comment]
> Did you forgot to change that?

Yes, unfortunately.

> My guess is that this behavior is not intended.

That's not, thats the weird interpretation of browsers on mobile phones about `:hover`, which is a single tap. I suspect. I can take that out with some CSS as well.

**Comment by loreanvictor** [comment]
fixes deployed via `0.2.12`

**Comment by bender2k14** [comment]
Both of the issue I mentioned in https://github.com/CONNECT-platform/codedoc/issues/57#issuecomment-674331066 are fix.  Great job.

I found one or two more potential / minor issues.

If I long press a word in a heading on mobile, then the word is highlighted, a menu appears, and I can select "Copy" to copy.  In the process, the link to the heading is not copied.  That is all as expected.

Here is the equivalent workflows on a desktop.  They are given in order of fastest to slowest.  Unfortunately, only the slowest one currently works.
1. Double left-click a word in a heading.  The word is temporarily highlighted but then is unhighlighted.  Also, the link to the heading is copied twice.
2. Click and drag across a word in a heading to highlight it.  Right-click to open the menu containing the option to copy.  The menu does not appear, the word is unhighlighted, and the link to the heading is copied.
3. Click and drag across a word in a heading to highlight it.  Use `Ctrl+C` to copy the highlighted word.

You did a great job differentiating touch interactions on mobile to distinguish copy link vs copy text behavior.  For desktop, can you similarly differentiate single- vs double-click to get (1) to work and left- vs right-click to get (2) to work?

**Comment by loreanvictor** [comment]
>  For desktop, can you similarly differentiate single- vs double-click to get (1) to work and left- vs right-click to get (2) to work?

Yes, both seem like good ideas which are easy to implement. However could you create a separate issue? The scope of this issue is actually being expanded beyond the original intent (which was confirmation behavior correction).

**Comment by bender2k14** [comment]
Sure.  See #65.

**Comment by loreanvictor** [comment]
perfect, thanks!
